server/sonar-server/src/test/java/org/sonar/server/es/queue/RecoveryIndexerTest.java

```diff
@@ -0,0 +1,397 @@
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestRule;
+import org.junit.rules.Timeout;
+import org.sonar.api.config.MapSettings;
+import org.sonar.api.config.Settings;
+import org.sonar.api.utils.internal.TestSystem2;
+import org.sonar.api.utils.log.LogTester;
+import org.sonar.api.utils.log.LoggerLevel;
+import org.sonar.db.DbSession;
+import org.sonar.db.DbTester;
+import org.sonar.db.es.EsQueueDto;
+import org.sonar.db.user.UserDto;
+import org.sonar.server.es.EsTester;
+import org.sonar.server.user.index.UserIndexDefinition;
+import org.sonar.server.user.index.UserIndexer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.sonar.api.utils.log.LoggerLevel.ERROR;
+import static org.sonar.api.utils.log.LoggerLevel.INFO;
+import static org.sonar.api.utils.log.LoggerLevel.TRACE;
+import static org.sonar.core.util.stream.MoreCollectors.toArrayList;
+
+public class RecoveryIndexerTest {
+
+  private static final long PAST = 1_000L;
+  private TestSystem2 system2 = new TestSystem2().setNow(PAST);
+
+  @Rule
+  public final EsTester es = new EsTester(new UserIndexDefinition(new MapSettings()));
+  @Rule
+  public final DbTester db = DbTester.create(system2);
+  @Rule
+  public final LogTester logTester = new LogTester().setLevel(TRACE);
+  @Rule
+  public TestRule safeguard = new Timeout(60, TimeUnit.SECONDS);
+
+  private RecoveryIndexer underTest;
+
+  @After
+  public void tearDown() {
+    if (underTest != null) {
+      underTest.stop();
+    }
+  }
+
+  @Test
+  public void display_default_configuration_at_startup() {
+    UserIndexer userIndexer = new UserIndexer(db.getDbClient(), es.client());
+    underTest = newRecoveryIndexer(userIndexer, new MapSettings());
+
+    underTest.start();
+
+    assertThat(logTester.logs(LoggerLevel.DEBUG)).contains(
+      "Elasticsearch recovery - sonar.search.recovery.delayInMs=300000",
+      "Elasticsearch recovery - sonar.search.recovery.minAgeInMs=300000");
+  }
+
+  @Test
+  public void start_triggers_recovery_run_at_fixed_rate() throws Exception {
+    Settings settings = new MapSettings()
+      .setProperty("sonar.search.recovery.initialDelayInMs", "0")
+      .setProperty("sonar.search.recovery.delayInMs", "1");
+    underTest = spy(new RecoveryIndexer(system2, settings, db.getDbClient(), mock(UserIndexer.class)));
+    AtomicInteger calls = new AtomicInteger(0);
+    doAnswer(invocation -> {
+      calls.incrementAndGet();
+      return null;
+    }).when(underTest).recover();
+
+    underTest.start();
+
+    // wait for 2 runs
+    while (calls.get() < 2) {
+      Thread.sleep(1L);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)